gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala

```diff
@@ -684,29 +754,54 @@ case class ColumnarOverrideRules(session: SparkSession)
     )
   }
 
-  override def preColumnarTransitions: Rule[SparkPlan] = plan =>
-    PhysicalPlanSelector.maybe(session, plan) {
-      setAdaptiveContext()
-      setOriginalPlan(plan)
-      transformPlan(preOverrides(), plan, "pre")
-    }
-
-  override def postColumnarTransitions: Rule[SparkPlan] = plan =>
-    PhysicalPlanSelector.maybe(session, plan) {
-      val planWithFallbackPolicy = transformPlan(fallbackPolicy(), plan, "fallback")
-      val finalPlan = planWithFallbackPolicy match {
-        case FallbackNode(fallbackPlan) =>
-          // we should use vanilla c2r rather than native c2r,
-          // and there should be no `GlutenPlan` any more,
-          // so skip the `postOverrides()`.
-          fallbackPlan
-        case plan =>
-          transformPlan(postOverrides(), plan, "post")
-      }
+  private def prepareFallback[T](plan: SparkPlan)(f: SparkPlan => T): T = {
+    setAdaptiveContext()
+    setOriginalPlan(plan)
+    try {
+      f(plan)
+    } finally {
       resetOriginalPlan()
       resetAdaptiveContext()
-      transformPlan(finallyRules(), finalPlan, "final")
     }
+  }
+
+  /**
+   * Note: Do not implement this API. We basically inject all of Gluten's physical rules through
+   * `postColumnarTransitions`.
+   *
+   * See: https://github.com/oap-project/gluten/pull/4790
+   */
+  final override def preColumnarTransitions: Rule[SparkPlan] = plan => {
+    // To infer caller's property: ApplyColumnarRulesAndInsertTransitions#outputsColumnar.
+    OutputsColumnarTester.wrap(plan)
+  }
+
+  override def postColumnarTransitions: Rule[SparkPlan] = plan => {
+    val outputsColumnar = OutputsColumnarTester.inferOutputsColumnar(plan)
```

Comment:
why not use `plan.supportsColumnar` ?

Reply:
Spark has `supportsRowBased` since 3.3 . A plan node can be both `supportsColumnar == true` and `supportsRowBased == true`.  We should know about the caller's intention having which property requested.

Reply:
If the plan support both columnar and row, then Spark should do thing. Does `plan.supportsColumnar` return different value with `OutputsColumnarTester.inferOutputsColumnar` ? It seems we never touch supportsRowBased.

Reply:
It's a rare case but it does covered by existing UTs...

Consider an `UnionExec` which supports both columnar and row-based execution, then even caller requested `outputsColumnar=false` through `ApplyColumnarRulesAndInsertTransitions#outputsColumnar`, we don't have a way to be aware of that since the `UnionExec` returns `true` for its `supportsColumnar` method.

So we should infer `outputsColumnar`'s exact value here rather than just calling `supportsColumnar`.

Reply:
Can we move OutputsColumnarTester related code to a new file with the comment ? The ColumnarOverrides file is too big.

Reply:
I was planning on splitting the file too. Let's do that in another patch. Thanks for the suggestion.